### ISSUE TITLE
Modifier keys and special keys are not handled correctly

### ISSUE THREAD

**Issue by chhackett** [chhackett/vty-windows#2]
As noted by others, function keys are not recognized properly.
Nor are holding modifier keys while pressing another key.
'BackTab' is not recognized (Shift + Tab).

All of the keyboard input data is available in the response to the ReadConsoleInputW.
However, the way it is processed, some data is ignored and some is just processed incorrectly.

Possible solutions:
- Improve handleInputEvent function such that keypresses that include special chars and/or modifier keys are processed immediately and pushed onto the eventChannel. This avoids the extra step of pushing bytes into the byte buffer, which has the problem that some keypresses don't have an easy way to translate into a byte sequence that can be interpreted correctly later on.
- Invent unique byte sequences that can unabiguously specify special char and modifier char sequences.

**Comment by ShrykeWindgrace** [comment]
- Function keys: I implemented that in the previous iteration, you can either cherrypick that code or I'll open a PR in a couple of days
- What do you understand under `modifier keys while pressing another key`? Combinations like `ctrl+d`? They work properly for me. Note that I filter out `\NUL` in `handleInputEvent` with a check `                    if isKeyDown && cwChar /= 0`
- `shift+tab` is indeed crashing (or force-exiting) my `vty-event-echo`, I'll look into that.

**Comment by ShrykeWindgrace** [comment]
Re: `shift-tab`
- it get sent as `\ESC[Z`
- ...and in out table the first match is just `\ESC`, hence `event-echo` exits
- the `mode-demo` exits on a `q`, there we can see that we receive three separate events `\ESC`, `[`, and `Z`

So all we need as a tactical solution is a sufficiently high-placed entry `("\ESC[Z",EvKey KBackTab [])` in our `ClassifyMap`. An equivalent solution would be to hardcode the capability `("kcbt", csi "Z")` in `windowsStringCaps`.

Strategically, I'd say we need to review the whole `ClassifyMap` - it surely contains a lot of redundant entries, as well as impossible ones.

**Comment by chhackett** [comment]
Agree about the classify map.
So did I cherry pick your F key work, and that works beautifully. Also added the backtab change, and that works too.

What I meant by the 'modifier keys while pressing another key' comment is that Alt + whatever didn't work, as well as some random additional cases. I fixed the Alt+key  for most keys. I'll have those changes checked in soon.

**Comment by chhackett** [comment]
So after playing with various modifier+key combos last night, I agree that we just need a better 'tactical' solution. I'll spend the next evening or two and really go through all possible key+modifier combos and verify that we're handling them all correctly, and fixing the remaining problems. (Some numpad combos still aren't working right).
I intend to clean up the generation of the ClassifyMap as I do this. I think there is room for improvement there.

**Comment by chhackett** [comment]
@ShrykeWindgrace 
I finished my work to fix all the known input key issues. I basically rewrote ANSIVT and cleaned it all up. Changes are checked into the keys-with-modifiers branch.

Now there are 3 categories of mappings, following directly from the microsoft input sequences page.

I verified that all keys + modifies worked except for combos that map to some application shortcut, or Windows shortcut...
like Ctrl+F11 makes a window fullscreen, etc, Alt+Tab switches windows, etc, etc etc. 

I'm wondering if we should filter out those cases from the mappings. But it doesn't really hurt anything - the application will just never receive those sequences.

Anyway, please take a look and let me know if you find any more room for improvement.

**Comment by ShrykeWindgrace** [comment]
That's great!

I do not think that we need to filter things like `alt+tab` out explicitly, that's not really surprising for an end user that such shortcut is handled by the OS. I have a very minor dislike for treating `shift+tab` as `BackTab` (vs leaving it as-is), but as long this behavior is clearly documented, I am ok with that. By the way, I'd like to take a look at how linux version handles this combination.
A bit surprised by `Ctrl+Z` becoming a `KPause` - I'd personally prefer leaving at as-is.

Another - major this time - problem is that I do not see combos like `C+S+something` at all; to my knowledge, windows does nothing fancy with them, my IntelliJ IDEA happily handles such shortcuts. https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#input-sequences is surprisiungly silent on that matter.

I did not manage to trigger a ``Ctrl+` `` event, even though  ``Alt+` `` works.

I'll try to take a look at this this week.

**Comment by ShrykeWindgrace** [comment]
Our current mapping has only these entries for `C+S+something`:

```
("\ESC[1;6A",EvKey KUp [MCtrl,MShift])
("\ESC[1;6B",EvKey KDown [MCtrl,MShift])
("\ESC[1;6C",EvKey KRight [MCtrl,MShift])
("\ESC[1;6D",EvKey KLeft [MCtrl,MShift])
("\ESC[1;6F",EvKey KEnd [MCtrl,MShift])
("\ESC[1;6H",EvKey KHome [MCtrl,MShift])
("\ESC[2;6~",EvKey KIns [MCtrl,MShift])
("\ESC[3;6~",EvKey KDel [MCtrl,MShift])
("\ESC[5;6~",EvKey KPageUp [MCtrl,MShift])
("\ESC[6;6~",EvKey KPageDown [MCtrl,MShift])
("\ESC[15;6~",EvKey (KFun 5) [MCtrl,MShift])
("\ESC[17;6~",EvKey (KFun 6) [MCtrl,MShift])
("\ESC[18;6~",EvKey (KFun 7) [MCtrl,MShift])
("\ESC[19;6~",EvKey (KFun 8) [MCtrl,MShift])
("\ESC[20;6~",EvKey (KFun 9) [MCtrl,MShift])
("\ESC[21;6~",EvKey (KFun 10) [MCtrl,MShift])
("\ESC[23;6~",EvKey (KFun 11) [MCtrl,MShift])
("\ESC[24;6~",EvKey (KFun 12) [MCtrl,MShift])
("\ESC[1;6P",EvKey (KFun 1) [MCtrl,MShift])
("\ESC[1;6Q",EvKey (KFun 2) [MCtrl,MShift])
("\ESC[1;6R",EvKey (KFun 3) [MCtrl,MShift])
("\ESC[1;6S",EvKey (KFun 4) [MCtrl,MShift])
```

**Comment by chhackett** [comment]
So I changed the 'Shift+Tab' mapping to EvKey KTab [MShift]. If that's the expected behavior, lets go with that.
I'm confused about your comment regarding Ctrl+Z mapping to KPause. 
When I test that, I get the correct result (testing with the mode-demo app): Last event: Just (EvKey (KChar 'z') [MCtrl])

Your last point - that Ctrl+Shift+anykey doesn't work. Yes, I think this is the main problem with the current approach of only using the VT sequences for processing input. 

Take, for example, various modifier combinations with the 'r' key:
- Ctrl + 'r':       "\DC2"
- Meta + 'r':     "\ESCr"
- Meta+Ctrl+'r': "\ESC\DC2"

These all are interpreted correctly. However, if you add Shift to any of these modifier combos, the resulting VT sequences are no different:
- Shift + Ctrl + 'r':    "\DC2"
etc...

So there is just no way, from VT sequences, we can ever tell the Shift key was pressed in combination with Ctrl or Meta...

However! We are not receiving just VT sequences - we have full KeyEventRecords. And we receive KeyEventRecords for each keypress, including when the user pressed the Shift, Ctrl keys. For example, in the case of pressing Shift then Ctrl then 'r', (then releasing those keys) we receive the following sequence of key events:

KeyEventRecordC {keyDown = True, repeatCount = 1, virtualKeyCode = 16, virtualScanCode = 42, uChar = 0, controlKeyStateK = 16}
KeyEventRecordC {keyDown = True, repeatCount = 1, virtualKeyCode = 17, virtualScanCode = 29, uChar = 0, controlKeyStateK = 24}
KeyEventRecordC {keyDown = True, repeatCount = 1, virtualKeyCode = 0, virtualScanCode = 0, uChar = 18, controlKeyStateK = 0}
KeyEventRecordC {keyDown = False, repeatCount = 1, virtualKeyCode = 82, virtualScanCode = 19, uChar = 18, controlKeyStateK = 24}
KeyEventRecordC {keyDown = False, repeatCount = 1, virtualKeyCode = 17, virtualScanCode = 29, uChar = 0, controlKeyStateK = 16}
KeyEventRecordC {keyDown = False, repeatCount = 1, virtualKeyCode = 16, virtualScanCode = 42, uChar = 0, controlKeyStateK = 0}

Notice that the first two records have virtualKeyCode of 16, then 17 respectively. Those indicate Shift is pressed, followed by Ctrl.

But we ignore those events because the uChar field is NUL. We only pass through the third event here, where uChar is 18 which results in the VT sequence of: "\DC2" which we translate to: EvKey (KChar 'r') [MCtrl]

So other apps are probably just processing the KeyEvent records directly, which would allows them to determine which modifier keys are pressed when the 'r' key is tapped.

So in conclusion, if we want to have the absolute best interpretation of modifier keys, we probably need to use the KeyEvent records directly to keep track of the state of modifier keys. This is what I was alluding to when I was suggesting strategies for solving this issue.

I think what I would recommend is we use the current VT sequence processing, but add a bit of extra state to the VT sequence output that indicates whether Shift is pressed as well. The Shift key seems to be the only bit missing from the VT sequences right now. (And in some cases, Shift is also handled correctly, as you pointed out for Function keys, and numpad keys, etc. 

This would make the input buffer code a fair bit more complex though. What do you think we should do?

**Comment by ShrykeWindgrace** [comment]
Thanks for the analysis!

I did some digging yesterday evening and arrived to the same conclusion. I still see different results for `Ctrl+Z`, but let's drop that for now.

Regarding my issue with `C+S+something` - I ran the tests on a linux machine, current master of `vty` does not see them either, so I propose to drop it for now. If in some distant future `conhost` implements `xterm`'s `modifyOtherKeys` or `CSI u` sequences, or invents its own protocol altogether, we could get back to the problem.

**Comment by chhackett** [comment]
That's funny - I just ran the same tests on Linux too, and saw same behavior as well. I guess we can drop it. I don't really fancy redesigning the whole input parsing framework right now. Maybe down the road a bit. For now, I'd like to just make sure we have parity with vty-unix, and release it.

Btw, do you want or need permissions to commit changes to this repo? If that's something you want or need, I'm happy to give that to you. I have never worked with other people on a github project before, so I'm not sure what's involved in having multiple people committing to a project.
